pkg/node/node.go

```diff
@@ -476,32 +524,105 @@ func (n *Node) monitorKeepPublicKeySubmission(
 	defer subscriptionConflictingPublicKey.Unsubscribe()
 	defer subscriptionPublicKeyPublished.Unsubscribe()
 
-	select {
-	case event := <-publicKeyPublished:
-		logger.Infof(
-			"public key [%x] has been accepted by keep: [%s]",
-			event.PublicKey,
-			keepAddress.String(),
-		)
-	case event := <-conflictingPublicKey:
+	startingBlock, err := n.ethereumChain.BlockCounter().CurrentBlock()
+	if err != nil {
 		logger.Errorf(
-			"member [%x] has submitted conflicting public key for keep [%s]: [%x]",
-			event.SubmittingMember,
+			"failed to get the starting block while setting up "+
+				"public key submission monitoring for keep [%s]: [%v]",
 			keepAddress.String(),
-			event.ConflictingPublicKey,
-		)
-	case <-monitoringCtx.Done():
-		logger.Warningf(
-			"monitoring of public key submission for keep [%s] "+
-				"has been cancelled: [%v]",
-			keepAddress.String(),
-			monitoringCtx.Err(),
+			err,
 		)
-	case <-abort:
-		logger.Warningf(
-			"monitoring of public key submission for keep [%s] "+
-				"has been aborted",
+		return
+	}
+
+	pubkeyCheckTrigger, err := n.ethereumChain.BlockCounter().BlockHeightWaiter(
+		startingBlock + blockConfirmations,
+	)
+	if err != nil {
+		logger.Errorf(
+			"failed to get the check trigger while setting up "+
+				"public key submission monitoring for keep [%s]: [%v]",
 			keepAddress.String(),
+			err,
 		)
+		return
+	}
+
+	for {
+		select {
+		case event := <-publicKeyPublished:
+			logger.Infof(
+				"public key [%x] has been accepted by keep: [%s]",
+				event.PublicKey,
+				keepAddress.String(),
+			)
```

Comment:
That's tricky but I think the 12 blocks confirmation that will potentially result in `return`ing from this `for`, should be triggered here. Otherwise, we are waiting for 12 blocks from the moment we submitted the transaction to the mempool and not from the moment it's been mined.

Reply:
We can use `WaitForBlockConfirmations` for that - if the function returns `true`, we can send a kill signal to a channel that will close this loop, e.g.

```
case event := <-publicKeyConfirmed:
  logger.Info("public key confirmed for keep [%v] [%v] blocks later, quitting monitoring", ...)
  return
case event := <- publicKeyPublished
  logger.Info("public key published for keep [%v], waiting [%v] blocks to confirm the state",...)

  isSignatureConfirmed, err := chainutil.WaitForBlockConfirmations()
  // if err, give up
  if (isSignatureConfirmed) {
    publicKeyConfirmed <-...
  }
case pubkeyCheckBlock := <-pubkeyCheckTrigger:
  // just like now, except that we do not return but just try to re-submit
```